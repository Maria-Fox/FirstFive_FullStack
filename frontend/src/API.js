import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "https://localhost:3001";

class API {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {
    // console.debug("API Call:", endpoint, data, method, this.token, "**********");

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}`, 'Access-Control-Allow-Origin': 'https://firstfive.onrender.comtps://firstfive.surge.sh'};
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      // console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };

  // ******************************************* Register/ LOGIN methods


  static async registerUser(userData) {
    let newUser = await this.request("auth/register", { ...userData }, 'post');
    this.token = newUser.signedJWT
    return newUser.signedJWT;
  };

  static async authenticateUser(userData) {
    let authUser = await this.request(`auth/login`, { ...userData }, 'post');
    return authUser.signedJWT;
  };




  // ******************************************* USER methods

  // Need to see about limmited views to only those who matched..


  static async viewUser(username) {
    let res = await this.request(`users/${username}`);
    return res.userData;
  };

  static async viewAuthUserProfile(username) {
    let res = await this.request(`users/${username}/profile`);
    return res.userData;
  };

  static async editUser(username, updateData) {
    let res = await this.request(`users/${username}`, { ...updateData }, 'patch');
    return res.userData;
  };

  static async deleteUser(username) {
    let res = await this.request(`users/${username}`, {}, 'delete');
    return res;
  };


  // ******************************************* PROJECT methods

  static async createProject(projectData) {
    let res = await this.request(`projects/add`, { ...projectData }, 'post');
    return res;
  };

  static async getAllProjects() {
    let res = await this.request(`projects/all`);
    return res;
  };

  static async getNonMatchedProjects() {
    let res = await this.request(`projects/unmatched`);
    return res;
  };

  static async carouselProjects() {
    let res = await this.request(`projects/carousel`);
    return res;
  };

  static async getUserCreatedProjects(username) {
    let res = await this.request(`projects/created/by/${username}`);
    return res;
  };

  static async viewProject(project_id) {
    let res = await this.request(`projects/${project_id}`);
    return res;
  };

  static async editProject(project_id, updateData) {
    let res = await this.request(`projects/${project_id}`, { ...updateData }, 'patch');
    return res;
  };

  static async deleteProject(project_id) {
    let res = await this.request(`projects/${project_id}`, {}, 'delete');
    return res;
  };

  // ******************************************* Match methods

  static async addMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'post');
    return res;
  };

  static async viewUserMatches(username) {
    let res = await this.request(`matches/${username}/all`);
    return res;
  };

  static async viewProjectUserMatches(project_id) {
    let res = await this.request(`matches/${project_id}/users`);
    return res;
  };

  static async removeUserMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'delete');
    return res;
  };


  // ******************************************* Project Member methods

  static async addProjectMember(project_id, username) {
    let res = await this.request(`projectmembers/add/${project_id}`, { username }, 'post');
    return res;
  };

  static async viewAllProjMembers(project_id) {
    let res = await this.request(`projectmembers/${project_id}/users`);
    return res;
  };

  static async deleteProjectMember(project_id, username) {
    let res = await this.request(`projectmembers/${project_id}`, { username }, 'delete');
    return res;
  };

  // ******************************************* Message methods

  static async createMessage(username, msgData) {
    let res = await this.request(`messages/${username}/create`, { ...msgData }, 'post');
    return res;
  };

  static async getAllUserMessages(username) {
    let res = await this.request(`messages/${username}/all`);
    return res;
  };

  static async viewSingleMsgData(message_id, username) {
    let res = await this.request(`messages/${username}/read/${message_id}`);
    return res;
  };


  // class closing bracket don't delete.
};

export default API;