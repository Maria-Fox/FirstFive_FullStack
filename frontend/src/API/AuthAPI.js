import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class AuthAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Register/ LOGIN methods


  static async registerUser(userData) {
    let newUser = await this.request("auth/register", { ...userData }, 'post');
    this.token = newUser.signedJWT
    return newUser.signedJWT;
  };

  static async authenticateUser(userData) {
    let authUser = await this.request(`auth/login`, { ...userData }, 'post');
    return authUser.signedJWT;
  };

};

export default AuthAPI;