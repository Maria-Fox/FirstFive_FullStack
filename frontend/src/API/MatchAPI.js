import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class MatchAPI {

  // Used to authenticate user. Sent in request header.
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Match methods

  static async addMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'post');
    return res;
  };

  static async viewUserMatches(username) {
    let res = await this.request(`matches/${username}/all`);
    return res;
  };

  static async removeUserMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'delete');
    return res;
  };

  static async viewProjectUserMatches(project_id) {
    let res = await this.request(`matches/${project_id}/users`);
    return res;
  };


};

export default MatchAPI;