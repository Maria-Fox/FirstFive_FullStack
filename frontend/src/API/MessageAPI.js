import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class MessageAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Message methods

  static async createMessage(username, msgData) {
    let res = await this.request(`messages/${username}/create`, { ...msgData }, 'post');
    return res;
  };

  static async getAllUserMessages(username) {
    let res = await this.request(`messages/${username}/all`);
    return res;
  };

  static async viewSingleMsgData(message_id, username) {
    let res = await this.request(`messages/${username}/read/${message_id}`);
    return res;
  };

};

export default MessageAPI;