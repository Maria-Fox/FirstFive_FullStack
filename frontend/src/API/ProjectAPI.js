import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class ProjectAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* PROJECT methods

  static async createProject(projectData) {
    let res = await this.request(`projects/add`, { ...projectData }, 'post');
    return res;
  };

  static async getAllProjects() {
    let res = await this.request(`projects/all`);
    return res;
  };

  static async getNonMatchedProjects() {
    let res = await this.request(`projects/unmatched`);
    return res;
  };

  static async carouselProjects() {
    let res = await this.request(`projects/carousel`);
    return res;
  };

  static async getUserCreatedProjects(username) {
    let res = await this.request(`projects/created/by/${username}`);
    return res;
  };

  static async viewProject(project_id) {
    let res = await this.request(`projects/${project_id}`);
    return res;
  };

  static async editProject(project_id, updateData) {
    let res = await this.request(`projects/${project_id}`, { ...updateData }, 'patch');
    return res;
  };

  static async deleteProject(project_id) {
    let res = await this.request(`projects/${project_id}`, {}, 'delete');
    return res;
  };

};

export default ProjectAPI;