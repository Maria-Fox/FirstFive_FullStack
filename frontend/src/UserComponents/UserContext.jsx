import React from "react";

// Create user context instance to later provide in App.js and use throughout application. Holds auth user

const UserContext = React.createContext();

export default UserContext;