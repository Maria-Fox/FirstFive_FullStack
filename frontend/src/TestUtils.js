// Used to create sample test user data & test UserContect provider..

import React from "react";
import UserContext from "./UserComponents/UserContext";

const authUser = {
  username: "softwareDev1"
};

// sample / data
const matchedProjectIds = [1, 2, 3, 4];

const userWithMatchState = { username: "softwareDev1", matchedProjectIds: [] };

// Creating a context provider and passing in the test user, matching ids. Passing in chidlren.
const UserProvider =
  ({ children, userWithMatchState }) => {
    <UserContext.Provider value={userWithMatchState}>
      {children}
    </UserContext.Provider>
  };

export { UserProvider, authUser, userWithMatchState };