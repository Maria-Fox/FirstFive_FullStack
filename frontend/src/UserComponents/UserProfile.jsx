import React, { useEffect, useState, useContext } from "react";
import { useParams, useNavigate } from "react-router-dom";
import API from "../API";
import UserContext from "./UserContext";
import AlertNotification from "../Common/AlertNotifications";
import { Card, Button } from "reactstrap";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {faVcard} from "@fortawesome/free-regular-svg-icons";

const UserProfile = () => {
  const { username } = useParams();
  const { authUser, setAuthUser } = useContext(UserContext);
  const [userData, setUserData] = useState(null);
  const [errors, setErrors] = useState(null);
  const navigate = useNavigate();

  // ***************************************************************

  useEffect(function userProfile() {
    async function viewUserProfile() {
      try {
        let user = await API.viewAuthUserProfile(username);
        setUserData({
          username: user.username,
          bio: user.bio,
          email: user.email
        });
      } catch (e) {
        setErrors(e);
        return;
      };
    };

    viewUserProfile();
  }, [setUserData, username]);

  // ***************************************************************


  // Need guidance on how I can update an owner - component wise.
  let navigateToUpdate = () => {
    navigate(`/users/update/${authUser}`);
  };

  // ***************************************************************


  let handleDelete = () => {
    try {
      API.deleteUser(authUser);
      setAuthUser(null);
      localStorage.removeItem("token");
      navigate("/auth/register");
    } catch (e) {
      setErrors(e);
      return;
    }
  };

  if (!userData) return <p>Loading...</p>

  // The routes are the same jus the type of route (get vs path is diff) how do I move around that?
  let profileOptions = (
    <div className="ProfileOptions">
      <Button onClick={navigateToUpdate}>Update Profile Details</Button>
      <div style={{ paddingTop: "20px" }}>
        <p style={{ color: "red" }}>Danger Zone</p>
        <Button outline color="danger" onClick={handleDelete} >
          Delete Profile</Button>
      </div>
    </div>
  );


  return (
    <div className="container" >
      {errors ? <AlertNotification messages={errors} /> : null}

      <h1 className="text-center text-white pt-2 mt-2"> 
            <FontAwesomeIcon icon={faVcard} />
            Profile</h1>


      <Card className="p-3 text-center pt-4 mt-2">

        <div >
          <h2>Username: {userData.username}</h2>
          <p>Bio: {userData.bio}</p>
          <p>Email: {userData.email}</p>

          {profileOptions}
        </div>
      </Card>
    </div>
  );
};

export default UserProfile;