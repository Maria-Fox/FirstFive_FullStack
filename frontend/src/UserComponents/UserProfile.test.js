import React from "react";
import { render, waitFor } from "@testing-library/react";
import UserProfile from "./UserProfile";
import UserContext from "./UserContext";
import { MemoryRouter } from "react-router-dom";
import { authUser } from "../TestUtils";
// Mock API
import API from "../API";
jest.mock("../API")

test("Smoke- Renders UserProfile component w/ valid user", function () {

  render(
    <MemoryRouter>
      <UserContext.Provider value={authUser}>
        <UserProfile />
      </UserContext.Provider>
    </MemoryRouter>
  );
});


test("Snapshot- Renders loading before API call.", function () {

  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={authUser}>
        <UserProfile />
      </UserContext.Provider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

test("Component has userData after API call", function () {

  jest.mock("axios");
  // Mock API call anad response data.

  API.mockImplementationOnce(() => {
    return {
      request: () => {
        return [
          {
            username: 1,
            email: "softwaredev1@gmail.com",
            bio: "Software Developer with almost a year of experience. Looking to work with a group using Node.js or React!",
          }
        ];
      },
    };
  });

  const utils = render(
    <UserContext.Provider value={authUser}>
      <MemoryRouter>
        <UserProfile />
      </MemoryRouter>
    </UserContext.Provider>
  );

  waitFor(() => {
    expect(utils.getByText("softwaredev1@gmail.com")).toBeInTheDocument();
    expect(utils.getByText("Software Developer with almost a year of experience. Looking to work with a group using Node.js or React!")).toBeInTheDocument();
  });
});