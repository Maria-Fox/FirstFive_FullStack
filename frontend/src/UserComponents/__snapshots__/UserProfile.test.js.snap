// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot- Renders loading before API call. 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;

exports[`Snapshot- Renders user profile for valid user 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
